db/dcrpg/upgrades.go

```diff
@@ -1195,3 +1227,74 @@ func versionAllTables(db *sql.DB, version TableVersion) error {
 	}
 	return nil
 }
+
+// verifyChainWork fetches and inserts missing chainwork values.
+// This addresses a table update done at DB version 3.7.0.
+func verifyChainWork(blockgate *rpcutils.BlockGate, db *sql.DB) (int64, error) {
+	// Count rows with missing chainWork.
+	var count int64
+	countRow := db.QueryRow(`SELECT COUNT(hash) FROM blocks WHERE chainwork IS NULL;`)
+	err := countRow.Scan(&count)
+	if err != nil {
+		log.Error("Failed to count null chainwork columns: %v", err)
+		return 0, err
+	}
+	if count == 0 {
+		return 0, nil
+	}
+
+	// Prepare the insertion statment. Parameters: 1. chainwork; 2. blockhash.
+	stmt, err := db.Prepare(`UPDATE blocks SET chainwork=$1 WHERE hash=$2;`)
+	if err != nil {
+		log.Error("Failed to prepare chainwork insertion statement: %v", err)
+		return 0, err
+	}
+	defer stmt.Close()
+
+	// Grab the blockhashes from rows that don't have chainwork.
+	rows, err := db.Query(`SELECT hash FROM blocks WHERE chainwork IS NULL;`)
+	if err != nil {
```

Comment:
You could avoid the `count` query by doing this query first, and then `if err == sql.ErrNoRows { return 0, nil }`, but since this code only runs once, it's no big deal.

Reply:
I use the count to calculate progress for the status reports also.

Reply:
Ah, I see that now.